File: packages/protocol/src/secure_transport.ts
import {
  createCipheriv,
  createDecipheriv,
  createHash,
  createPrivateKey,
  createPublicKey,
  diffieHellman,
  generateKeyPairSync,
  hkdfSync,
  randomUUID,
  sign,
  timingSafeEqual,
  verify,
  type KeyObject,
} from "node:crypto";
import { canonicalizeJson, verifyCredentialSignature, type SignedCredential } from "./pairing.js";

/**
 * End-to-end encryption for the bridge transport.
 *
 * The relay forwards opaque payloads but terminates TLS, so a relay operator or
 * anyone who compromises that host can read task titles and message text in
 * flight. This layer removes the relay from the trust boundary: each connection
 * agrees a fresh key directly between the computer and the paired device.
 *
 * Ephemeral X25519 keys give forward secrecy, and each side signs its ephemeral
 * key with the Ed25519 identity it already owns from pairing — the host with its
 * host identity, the device with the key inside its host-signed credential. No
 * new pairing material is required, so existing pairings keep working.
 *
 * Routing fields stay in clear text because the relay needs them; only the
 * bridge envelope itself is encrypted.
 */

export const SECURE_TRANSPORT_SCHEME = "x25519-hkdf-sha256-aes256gcm" as const;
export const SECURE_TRANSPORT_VERSION = 1 as const;

const HKDF_HASH = "sha256";
const KEY_BYTES = 32;
const NONCE_BYTES = 12;
const TAG_BYTES = 16;
const RAW_X25519_BYTES = 32;
/** DER prefix for an X25519 SubjectPublicKeyInfo, so raw 32-byte keys can cross the wire. */
const X25519_SPKI_PREFIX = Buffer.from("302a300506032b656e032100", "hex");
/** AES-GCM must never repeat a (key, nonce) pair; a session is closed long before this. */
const MAX_FRAME_COUNTER = 2 ** 48;
/** One decrypt allocation ceiling, matched to the relay's own payload limit. */
export const MAX_SECURE_FRAME_BYTES = 8 * 1024 * 1024;

export type SecureTransportRole = "host" | "device";

export interface SecureHandshakeOffer {
  readonly scheme: typeof SECURE_TRANSPORT_SCHEME;
  readonly version: typeof SECURE_TRANSPORT_VERSION;
  /** Base64url raw 32-byte X25519 public key, fresh for this connection. */
  readonly ephemeralPublicKey: string;
  /** Base64url Ed25519 signature over the canonical bind for this offer. */
  readonly signature: string;
}

export interface SecureHandshakeAccept extends SecureHandshakeOffer {
  readonly kind: "secure_handshake";
  /** The host-signed credential naming the device key that signed this accept. */
  readonly credential: SignedCredential;
  readonly deviceId: string;
}

export interface SecureFrame {
  readonly kind: "secure";
  /** Frame counter for this direction, used as the AES-GCM nonce and as AAD. */
  readonly counter: number;
  /** Base64url ciphertext with its appended GCM tag. */
  readonly ciphertext: string;
}

function base64url(input: Uint8Array): string {
  return Buffer.from(input).toString("base64url");
}

function fromBase64url(input: string): Buffer {
  return Buffer.from(input, "base64url");
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

export function rawX25519PublicKey(key: KeyObject): Buffer {
  const spki = key.export({ type: "spki", format: "der" });
  return Buffer.from(spki.subarray(spki.length - RAW_X25519_BYTES));
}

export function importX25519PublicKey(raw: Buffer): KeyObject {
  if (raw.length !== RAW_X25519_BYTES) throw new Error("An X25519 public key must be 32 bytes");
  return createPublicKey({ key: Buffer.concat([X25519_SPKI_PREFIX, raw]), format: "der", type: "spki" });
}

export interface EphemeralKeyPair {
  readonly publicKey: KeyObject;
  readonly privateKey: KeyObject;
}

export function createEphemeralKeyPair(): EphemeralKeyPair {
  return generateKeyPairSync("x25519");
}

/**
 * The exact bytes each side signs. Every value that identifies the session is
 * covered, so a relay cannot substitute its own ephemeral key: the device's
 * signature commits to the host key it actually received, and the host's
 * signature commits to the key it actually sent.
 */
export function handshakeBind(input: {
  readonly role: SecureTransportRole;
  readonly hostId: string;
  readonly deviceId?: string;
  readonly hostEphemeralPublicKey: string;
  readonly deviceEphemeralPublicKey?: string;
}): Buffer {
  return Buffer.from(canonicalizeJson({
    purpose: "tethoq-secure-transport-handshake",
    scheme: SECURE_TRANSPORT_SCHEME,
    version: SECURE_TRANSPORT_VERSION,
    role: input.role,
    hostId: input.hostId,
    deviceId: input.deviceId ?? null,
    hostEphemeralPublicKey: input.hostEphemeralPublicKey,
    deviceEphemeralPublicKey: input.deviceEphemeralPublicKey ?? null,
  }), "utf8");
}

export interface SecureTransportKeys {
  readonly hostToDevice: Buffer;
  readonly deviceToHost: Buffer;
}

/**
 * Both ephemeral public keys are folded into the HKDF salt, so a session key is
 * only reachable by the two parties that chose those keys. Each direction gets
 * its own key so a reflected frame can never decrypt.
 */
export function deriveSecureTransportKeys(input: {
  readonly sharedSecret: Buffer;
  readonly hostId: string;
  readonly deviceId: string;
  readonly hostEphemeralPublicKey: string;
  readonly deviceEphemeralPublicKey: string;
}): SecureTransportKeys {
  if (input.sharedSecret.length !== KEY_BYTES) throw new Error("The shared secret must be 32 bytes");
  const salt = createHash(HKDF_HASH).update(canonicalizeJson({
    purpose: "tethoq-secure-transport-salt",
    scheme: SECURE_TRANSPORT_SCHEME,
    version: SECURE_TRANSPORT_VERSION,
    hostId: input.hostId,
    deviceId: input.deviceId,
    hostEphemeralPublicKey: input.hostEphemeralPublicKey,
    deviceEphemeralPublicKey: input.deviceEphemeralPublicKey,
  })).digest();
  const derive = (label: string): Buffer =>
    Buffer.from(hkdfSync(HKDF_HASH, input.sharedSecret, salt, Buffer.from(label, "utf8"), KEY_BYTES));
  return {
    hostToDevice: derive("tethoq-secure-transport-v1:host-to-device"),
    deviceToHost: derive("tethoq-secure-transport-v1:device-to-host"),
  };
}

export function secureSharedSecret(privateKey: KeyObject, peerPublicKey: KeyObject): Buffer {
  return diffieHellman({ privateKey, publicKey: peerPublicKey });
}

function nonce(counter: number): Buffer {
  if (!Number.isInteger(counter) || counter < 0 || counter >= MAX_FRAME_COUNTER) {
    throw new Error("The secure transport frame counter is out of range");
  }
  const value = Buffer.alloc(NONCE_BYTES);
  value.writeUIntBE(counter, NONCE_BYTES - 6, 6);
  return value;
}

function associatedData(direction: string, counter: number): Buffer {
  return Buffer.from(`${SECURE_TRANSPORT_SCHEME}:${SECURE_TRANSPORT_VERSION}:${direction}:${counter}`, "utf8");
}

export function sealSecureFrame(key: Buffer, direction: string, counter: number, plaintext: string): SecureFrame {
  const cipher = createCipheriv("aes-256-gcm", key, nonce(counter));
  cipher.setAAD(associatedData(direction, counter));
  const body = Buffer.concat([cipher.update(plaintext, "utf8"), cipher.final()]);
  return { kind: "secure", counter, ciphertext: base64url(Buffer.concat([body, cipher.getAuthTag()])) };
}

export function openSecureFrame(key: Buffer, direction: string, frame: SecureFrame): string {
  const sealed = fromBase64url(frame.ciphertext);
  if (sealed.length < TAG_BYTES) throw new Error("The secure frame is too short to authenticate");
  if (sealed.length > MAX_SECURE_FRAME_BYTES) throw new Error("The secure frame is too large");
  const decipher = createDecipheriv("aes-256-gcm", key, nonce(frame.counter));
  decipher.setAAD(associatedData(direction, frame.counter));
  decipher.setAuthTag(sealed.subarray(sealed.length - TAG_BYTES));
  return Buffer.concat([decipher.update(sealed.subarray(0, sealed.length - TAG_BYTES)), decipher.final()]).toString("utf8");
}

export function parseSecureFrame(value: unknown): SecureFrame | null {
  if (!isRecord(value) || value.kind !== "secure") return null;
  if (typeof value.counter !== "number" || !Number.isInteger(value.counter) || value.counter < 0) {
    throw new Error("The secure frame counter is invalid");
  }
  if (typeof value.ciphertext !== "string" || value.ciphertext.length === 0) {
    throw new Error("The secure frame ciphertext is invalid");
  }
  return { kind: "secure", counter: value.counter, ciphertext: value.ciphertext };
}

export function parseSecureHandshakeOffer(value: unknown): SecureHandshakeOffer | null {
  if (!isRecord(value)) return null;
  if (value.scheme !== SECURE_TRANSPORT_SCHEME || value.version !== SECURE_TRANSPORT_VERSION) return null;
  if (typeof value.ephemeralPublicKey !== "string" || typeof value.signature !== "string") return null;
  return {
    scheme: SECURE_TRANSPORT_SCHEME,
    version: SECURE_TRANSPORT_VERSION,
    ephemeralPublicKey: value.ephemeralPublicKey,
    signature: value.signature,
  };
}

export function parseSecureHandshakeAccept(value: unknown): SecureHandshakeAccept | null {
  if (!isRecord(value) || value.kind !== "secure_handshake") return null;
  const offer = parseSecureHandshakeOffer(value);
  if (offer === null) throw new Error("The secure handshake scheme is not supported");
  if (!isRecord(value.credential) || typeof value.credential.payload !== "string" || typeof value.credential.signature !== "string") {
    throw new Error("The secure handshake credential is malformed");
  }
  if (typeof value.deviceId !== "string" || value.deviceId.length === 0) throw new Error("The secure handshake device is missing");
  return {
    ...offer,
    kind: "secure_handshake",
    deviceId: value.deviceId,
    credential: { payload: value.credential.payload, signature: value.credential.signature },
  };
}

/** Host side: the offer sent inside `protocol.hello`. */
export function createSecureHandshakeOffer(input: {
  readonly hostId: string;
  readonly hostPrivateKeyPem: string;
  readonly keyPair?: EphemeralKeyPair;
}): { readonly offer: SecureHandshakeOffer; readonly keyPair: EphemeralKeyPair } {
  const keyPair = input.keyPair ?? createEphemeralKeyPair();
  const ephemeralPublicKey = base64url(rawX25519PublicKey(keyPair.publicKey));
  const signature = sign(null, handshakeBind({
    role: "host",
    hostId: input.hostId,
    hostEphemeralPublicKey: ephemeralPublicKey,
  }), createPrivateKey(input.hostPrivateKeyPem));
  return {
    offer: { scheme: SECURE_TRANSPORT_SCHEME, version: SECURE_TRANSPORT_VERSION, ephemeralPublicKey, signature: base64url(signature) },
    keyPair,
  };
}

/**
 * Device side: verifies the host's offer against the host key stored at pairing,
 * then answers with its own signed ephemeral key.
 */
export function acceptSecureHandshake(input: {
  readonly offer: SecureHandshakeOffer;
  readonly hostId: string;
  readonly hostPublicKeyPem: string;
  readonly deviceId: string;
  readonly devicePrivateKeyPem: string;
  readonly credential: SignedCredential;
  readonly keyPair?: EphemeralKeyPair;
}): { readonly accept: SecureHandshakeAccept; readonly keys: SecureTransportKeys } {
  const hostEphemeralPublicKey = input.offer.ephemeralPublicKey;
  const hostOfferValid = verify(
    null,
    handshakeBind({ role: "host", hostId: input.hostId, hostEphemeralPublicKey }),
    createPublicKey(input.hostPublicKeyPem),
    fromBase64url(input.offer.signature),
  );
  if (!hostOfferValid) throw new Error("The computer's encryption key was not signed by the paired host");
  const keyPair = input.keyPair ?? createEphemeralKeyPair();
  const deviceEphemeralPublicKey = base64url(rawX25519PublicKey(keyPair.publicKey));
  const signature = sign(null, handshakeBind({
    role: "device",
    hostId: input.hostId,
    deviceId: input.deviceId,
    hostEphemeralPublicKey,
    deviceEphemeralPublicKey,
  }), createPrivateKey(input.devicePrivateKeyPem));
  const keys = deriveSecureTransportKeys({
    sharedSecret: secureSharedSecret(keyPair.privateKey, importX25519PublicKey(fromBase64url(hostEphemeralPublicKey))),
    hostId: input.hostId,
    deviceId: input.deviceId,
    hostEphemeralPublicKey,
    deviceEphemeralPublicKey,
  });
  return {
    accept: {
      kind: "secure_handshake",
      scheme: SECURE_TRANSPORT_SCHEME,
      version: SECURE_TRANSPORT_VERSION,
      deviceId: input.deviceId,
      credential: input.credential,
      ephemeralPublicKey: deviceEphemeralPublicKey,
      signature: base64url(signature),
    },
    keys,
  };
}

/**
 * Host side: verifies the device's accept and derives the same keys. The
 * credential is checked by the caller's pairing manager first, which is what
 * proves the signing key belongs to a device this host actually paired.
 */
export function completeSecureHandshake(input: {
  readonly accept: SecureHandshakeAccept;
  readonly hostId: string;
  readonly hostPublicKeyPem: string;
  readonly keyPair: EphemeralKeyPair;
}): { readonly keys: SecureTransportKeys; readonly deviceId: string; readonly credentialId: string } {
  const credential = verifyCredentialSignature(input.accept.credential, input.hostPublicKeyPem);
  if (credential.hostId !== input.hostId) throw new Error("The secure handshake credential belongs to another host");
  if (credential.deviceId !== input.accept.deviceId) throw new Error("The secure handshake device does not match its credential");
  const hostEphemeralPublicKey = base64url(rawX25519PublicKey(input.keyPair.publicKey));
  const deviceEphemeralPublicKey = input.accept.ephemeralPublicKey;
  const valid = verify(
    null,
    handshakeBind({
      role: "device",
      hostId: input.hostId,
      deviceId: credential.deviceId,
      hostEphemeralPublicKey,
      deviceEphemeralPublicKey,
    }),
    createPublicKey(credential.devicePublicKeyPem),
    fromBase64url(input.accept.signature),
  );
  if (!valid) throw new Error("The device's encryption key signature is invalid");
  const keys = deriveSecureTransportKeys({
    sharedSecret: secureSharedSecret(input.keyPair.privateKey, importX25519PublicKey(fromBase64url(deviceEphemeralPublicKey))),
    hostId: input.hostId,
    deviceId: credential.deviceId,
    hostEphemeralPublicKey,
    deviceEphemeralPublicKey,
  });
  return { keys, deviceId: credential.deviceId, credentialId: credential.credentialId };
}

/**
 * One direction of an established session. Outbound counters increase; inbound
 * counters must not repeat or move backwards, which rejects a replayed frame
 * even though the relay can reorder or duplicate what it forwards.
 */
export class SecureChannel {
  #outboundCounter = 0;
  #highestInboundCounter = -1;

  public constructor(
    private readonly keys: SecureTransportKeys,
    private readonly role: SecureTransportRole,
    public readonly sessionId: string = randomUUID(),
  ) {}

  get #sendKey(): Buffer {
    return this.role === "host" ? this.keys.hostToDevice : this.keys.deviceToHost;
  }

  get #receiveKey(): Buffer {
    return this.role === "host" ? this.keys.deviceToHost : this.keys.hostToDevice;
  }

  get #sendDirection(): string {
    return this.role === "host" ? "host-to-device" : "device-to-host";
  }

  get #receiveDirection(): string {
    return this.role === "host" ? "device-to-host" : "host-to-device";
  }

  public seal(plaintext: string): SecureFrame {
    const frame = sealSecureFrame(this.#sendKey, this.#sendDirection, this.#outboundCounter, plaintext);
    this.#outboundCounter += 1;
    return frame;
  }

  public open(frame: SecureFrame): string {
    if (frame.counter <= this.#highestInboundCounter) throw new Error("A secure transport frame was replayed or reordered");
    const plaintext = openSecureFrame(this.#receiveKey, this.#receiveDirection, frame);
    this.#highestInboundCounter = frame.counter;
    return plaintext;
  }

  /**
   * Short human-comparable fingerprint of the agreed keys. Both ends print the
   * same value only when no one sat in the middle.
   */
  public fingerprint(): string {
    const digest = createHash("sha256")
      .update(this.keys.hostToDevice)
      .update(this.keys.deviceToHost)
      .digest();
    return [...digest.subarray(0, 4)].map((byte) => byte.toString(16).padStart(2, "0")).join("");
  }

  public matches(other: SecureChannel): boolean {
    const left = Buffer.from(this.fingerprint(), "hex");
    const right = Buffer.from(other.fingerprint(), "hex");
    return left.length === right.length && timingSafeEqual(left, right);
  }
}

File: apps/remote_client/lib/src/secure_transport.dart
import 'dart:convert';
import 'dart:typed_data';

import 'package:cryptography/cryptography.dart';

import 'json.dart';
import 'models.dart';
import 'security.dart';

/// Device half of the bridge transport encryption.
///
/// The relay forwards messages but terminates TLS, so without this layer a relay
/// operator can read task titles and message text in flight. Each connection
/// agrees a fresh key straight with the paired computer instead.
///
/// Every constant, label, and canonical string here must match
/// `packages/protocol/src/secure_transport.ts` byte for byte; the shared vectors
/// in `test/secure_transport_test.dart` pin that agreement.
const String secureTransportScheme = 'x25519-hkdf-sha256-aes256gcm';
const int secureTransportVersion = 1;

const int _keyBytes = 32;
const int _nonceBytes = 12;
const int _macBytes = 16;
const String _hostToDevice = 'host-to-device';
const String _deviceToHost = 'device-to-host';

final X25519 _x25519 = X25519();
final Ed25519 _ed25519 = Ed25519();
final Hkdf _hkdf = Hkdf(hmac: Hmac.sha256(), outputLength: _keyBytes);
final AesGcm _aes = AesGcm.with256bits(nonceLength: _nonceBytes);

class SecureTransportException implements Exception {
  const SecureTransportException(this.message);
  final String message;
  @override
  String toString() => message;
}

/// The host's signed offer, carried inside `protocol.hello`.
class SecureHandshakeOffer {
  const SecureHandshakeOffer(
      {required this.ephemeralPublicKey, required this.signature});

  /// Returns null when the computer did not offer encryption at all, which is
  /// how an older bridge behaves.
  static SecureHandshakeOffer? tryParse(Object? value) {
    if (value is! Map<Object?, Object?>) return null;
    final json = jsonMap(value, name: 'secure handshake offer');
    if (json['scheme'] != secureTransportScheme ||
        json['version'] != secureTransportVersion) {
      throw const SecureTransportException(
          'This computer offered an encryption scheme this app does not support');
    }
    return SecureHandshakeOffer(
      ephemeralPublicKey: requireString(json, 'ephemeralPublicKey'),
      signature: requireString(json, 'signature'),
    );
  }

  final String ephemeralPublicKey;
  final String signature;
}

/// The exact bytes each side signs. Mirrors `handshakeBind` in the TypeScript
/// module: every value that identifies the session is covered, so a relay
/// cannot substitute its own ephemeral key without breaking a signature.
Uint8List handshakeBind({
  required String role,
  required String hostId,
  String? deviceId,
  required String hostEphemeralPublicKey,
  String? deviceEphemeralPublicKey,
}) {
  return Uint8List.fromList(utf8.encode(canonicalJson(<String, Object?>{
    'purpose': 'tethoq-secure-transport-handshake',
    'scheme': secureTransportScheme,
    'version': secureTransportVersion,
    'role': role,
    'hostId': hostId,
    'deviceId': deviceId,
    'hostEphemeralPublicKey': hostEphemeralPublicKey,
    'deviceEphemeralPublicKey': deviceEphemeralPublicKey,
  })));
}

class SecureTransportKeys {
  const SecureTransportKeys(
      {required this.hostToDevice, required this.deviceToHost});
  final List<int> hostToDevice;
  final List<int> deviceToHost;
}

/// Both ephemeral keys are folded into the HKDF salt, so the session key is
/// reachable only by the two parties that chose them. Each direction gets its
/// own key, so a reflected frame can never decrypt.
Future<SecureTransportKeys> deriveSecureTransportKeys({
  required List<int> sharedSecret,
  required String hostId,
  required String deviceId,
  required String hostEphemeralPublicKey,
  required String deviceEphemeralPublicKey,
}) async {
  if (sharedSecret.length != _keyBytes) {
    throw const SecureTransportException('The shared secret must be 32 bytes');
  }
  final salt = (await Sha256().hash(utf8.encode(canonicalJson(<String, Object?>{
    'purpose': 'tethoq-secure-transport-salt',
    'scheme': secureTransportScheme,
    'version': secureTransportVersion,
    'hostId': hostId,
    'deviceId': deviceId,
    'hostEphemeralPublicKey': hostEphemeralPublicKey,
    'deviceEphemeralPublicKey': deviceEphemeralPublicKey,
  }))))
      .bytes;
  Future<List<int>> derive(String label) async {
    final key = await _hkdf.deriveKey(
      secretKey: SecretKey(sharedSecret),
      nonce: salt,
      info: utf8.encode(label),
    );
    return key.extractBytes();
  }

  return SecureTransportKeys(
    hostToDevice: await derive('tethoq-secure-transport-v1:host-to-device'),
    deviceToHost: await derive('tethoq-secure-transport-v1:device-to-host'),
  );
}

Uint8List _nonce(int counter) {
  if (counter < 0 || counter >= 0x1000000000000) {
    throw const SecureTransportException(
        'The secure transport frame counter is out of range');
  }
  final value = Uint8List(_nonceBytes);
  var remaining = counter;
  for (var index = _nonceBytes - 1; index >= _nonceBytes - 6; index -= 1) {
    value[index] = remaining & 0xff;
    remaining = remaining >> 8;
  }
  return value;
}

List<int> _associatedData(String direction, int counter) => utf8
    .encode('$secureTransportScheme:$secureTransportVersion:$direction:$counter');

/// One established session. Outbound counters increase; an inbound counter must
/// move forward, which rejects a frame the relay replayed or reordered.
class SecureChannel {
  SecureChannel(this._keys);

  final SecureTransportKeys _keys;
  int _outboundCounter = 0;
  int _highestInboundCounter = -1;

  Future<JsonMap> seal(String plaintext) async {
    final counter = _outboundCounter;
    _outboundCounter += 1;
    final box = await _aes.encrypt(
      utf8.encode(plaintext),
      secretKey: SecretKey(_keys.deviceToHost),
      nonce: _nonce(counter),
      aad: _associatedData(_deviceToHost, counter),
    );
    return <String, Object?>{
      'kind': 'secure',
      'counter': counter,
      'ciphertext':
          base64UrlNoPadding(<int>[...box.cipherText, ...box.mac.bytes]),
    };
  }

  Future<String> open(JsonMap frame) async {
    final counter = frame['counter'];
    if (counter is! int || counter < 0) {
      throw const SecureTransportException(
          'The secure frame counter is invalid');
    }
    if (counter <= _highestInboundCounter) {
      throw const SecureTransportException(
          'A secure transport frame was replayed or reordered');
    }
    final sealed = decodeBase64Url(requireString(frame, 'ciphertext'));
    if (sealed.length < _macBytes) {
      throw const SecureTransportException(
          'The secure frame is too short to authenticate');
    }
    final plaintext = await _aes.decrypt(
      SecretBox(
        sealed.sublist(0, sealed.length - _macBytes),
        nonce: _nonce(counter),
        mac: Mac(sealed.sublist(sealed.length - _macBytes)),
      ),
      secretKey: SecretKey(_keys.hostToDevice),
      aad: _associatedData(_hostToDevice, counter),
    );
    _highestInboundCounter = counter;
    return utf8.decode(plaintext);
  }

  /// Short human-comparable fingerprint. Both ends print the same value only
  /// when no one sat in the middle.
  Future<String> fingerprint() async {
    final digest = await Sha256()
        .hash(<int>[..._keys.hostToDevice, ..._keys.deviceToHost]);
    return digest.bytes
        .sublist(0, 4)
        .map((byte) => byte.toRadixString(16).padLeft(2, '0'))
        .join();
  }
}

/// Proves to the relay that this device is the one its host-signed credential
/// names. The room token is shared with every paired device, so without this any
/// of them could claim another device's ID, evict its tunnel, and receive
/// everything addressed to it. The relay already pinned the computer's public
/// key, so it can check the credential without any new key exchange.
Future<JsonMap> signRelayDeviceAttach({
  required PairedHost host,
  required String token,
}) async {
  final attachId = randomId('attach');
  final issuedAt = DateTime.now().toUtc().toIso8601String();
  final tokenDigest = (await Sha256().hash(utf8.encode(token)))
      .bytes
      .map((byte) => byte.toRadixString(16).padLeft(2, '0'))
      .join();
  final signature = await _ed25519.sign(
    utf8.encode(canonicalJson(<String, Object?>{
      'purpose': 'tethoq-relay-device-attach',
      'version': 1,
      'hostId': host.hostId,
      'deviceId': host.deviceId,
      'attachId': attachId,
      'issuedAt': issuedAt,
      'tokenDigest': tokenDigest,
    })),
    keyPair: SimpleKeyPairData(
      host.devicePrivateKey,
      publicKey:
          SimplePublicKey(host.devicePublicKey, type: KeyPairType.ed25519),
      type: KeyPairType.ed25519,
    ),
  );
  return <String, Object?>{
    'version': 1,
    'credential': host.credential.toJson(),
    'attachId': attachId,
    'issuedAt': issuedAt,
    'signature': base64UrlNoPadding(signature.bytes),
  };
}

class SecureHandshakeResult {
  const SecureHandshakeResult({required this.accept, required this.channel});
  final JsonMap accept;
  final SecureChannel channel;
}

/// Verifies the computer's offer against the host key stored at pairing, then
/// answers with this device's own signed ephemeral key.
Future<SecureHandshakeResult> acceptSecureHandshake({
  required SecureHandshakeOffer offer,
  required PairedHost host,
}) async {
  final hostSignatureValid = await _ed25519.verify(
    handshakeBind(
      role: 'host',
      hostId: host.hostId,
      hostEphemeralPublicKey: offer.ephemeralPublicKey,
    ),
    signature: Signature(
      decodeBase64Url(offer.signature),
      publicKey: SimplePublicKey(
          ed25519RawPublicKeyFromPem(host.hostPublicKeyPem),
          type: KeyPairType.ed25519),
    ),
  );
  if (!hostSignatureValid) {
    throw const SecureTransportException(
        'This computer\'s encryption key was not signed by the paired host');
  }

  final ephemeral = await _x25519.newKeyPair();
  final deviceEphemeralPublicKey =
      base64UrlNoPadding((await ephemeral.extractPublicKey()).bytes);
  final signature = await _ed25519.sign(
    handshakeBind(
      role: 'device',
      hostId: host.hostId,
      deviceId: host.deviceId,
      hostEphemeralPublicKey: offer.ephemeralPublicKey,
      deviceEphemeralPublicKey: deviceEphemeralPublicKey,
    ),
    keyPair: SimpleKeyPairData(
      host.devicePrivateKey,
      publicKey:
          SimplePublicKey(host.devicePublicKey, type: KeyPairType.ed25519),
      type: KeyPairType.ed25519,
    ),
  );
  final shared = await _x25519.sharedSecretKey(
    keyPair: ephemeral,
    remotePublicKey: SimplePublicKey(
        decodeBase64Url(offer.ephemeralPublicKey),
        type: KeyPairType.x25519),
  );
  final keys = await deriveSecureTransportKeys(
    sharedSecret: await shared.extractBytes(),
    hostId: host.hostId,
    deviceId: host.deviceId,
    hostEphemeralPublicKey: offer.ephemeralPublicKey,
    deviceEphemeralPublicKey: deviceEphemeralPublicKey,
  );
  return SecureHandshakeResult(
    accept: <String, Object?>{
      'kind': 'secure_handshake',
      'scheme': secureTransportScheme,
      'version': secureTransportVersion,
      'deviceId': host.deviceId,
      'credential': host.credential.toJson(),
      'ephemeralPublicKey': deviceEphemeralPublicKey,
      'signature': base64UrlNoPadding(signature.bytes),
    },
    channel: SecureChannel(keys),
  );
}
